### PR TITLE
review adjuster class: reduce memory allocations/deallocations, also …

### PR DESCRIPTION
**Pull request by heckflosse**
…small speedup for adjuster creation

### CODE REVIEW

**Review by Hombre57 — APPROVED** [review]
Didn't had time to build and test, but by looking at the diff, it looks good to me, for what I can tell of course. Thanks for the cleanup.

File: rtgui/adjuster.cc [hunk 1]
```diff
@@ -27,91 +27,90 @@
...
+namespace {
+double one2one(double val)
 {
     return val;
 }
+}
 
-Adjuster::Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep, double vdefault, Gtk::Image *imgIcon1, Gtk::Image *imgIcon2, double2double_fun slider2value_, double2double_fun value2slider_)
-{
+Adjuster::Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep, double vdefault, Gtk::Image *imgIcon1, Gtk::Image *imgIcon2, double2double_fun slider2value_, double2double_fun value2slider_) :
```
Comment (Floessie) [inline rtgui/adjuster.cc]: How about breaking this line, too?

File: rtgui/adjuster.cc [hunk 2]
```diff
@@ -27,91 +27,90 @@
...
+    blocked(false),
+    addMode(false),
+    eventPending(false),
+    vMin(vmin),
+    vMax(vmax),
+    vStep(vstep),
+    logBase(0),
+    logPivot(0),
+    logAnchorMiddle(false),
+    value2slider(value2slider_ ? value2slider_ : one2one),
```
Comment (Floessie) [inline rtgui/adjuster.cc]: `&one2one` makes is more readable.

If those function pointers were `std::function<>`s one could provide a function with context (closure). Don't know if that might once be needed...

File: rtgui/adjuster.cc [hunk 3]
```diff
@@ -27,91 +27,90 @@
...
-    afterReset = false;
-    blocked = false;
-    automatic = nullptr;
-    eventPending = false;
-    grid = NULL;
-    imageIcon1 = imgIcon1;
-
-    logBase = 0;
-    logPivot = 0;
-    logAnchorMiddle = false;
```
Comment (Floessie) [inline rtgui/adjuster.cc:51]: Thank you! :+1:

File: rtgui/adjuster.cc [hunk 4]
```diff
@@ -27,91 +27,90 @@
 
 #define MIN_RESET_BUTTON_HEIGHT 17
```
Comment (Floessie) [inline rtgui/adjuster.cc]: `constexpr`.

File: rtgui/adjuster.cc [hunk 5]
```diff
@@ -129,47 +128,41 @@ Adjuster::Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep
...
-    delayConnection.block (true);
+    sliderChange.block(true);
+    spinChange.block(true);
+    delayConnection.block(true);
     adjusterListener = nullptr;
 
-    if (automatic) {
-        delete automatic;
-    }
+    delete automatic;
```
Comment (Floessie) [inline rtgui/adjuster.cc]: Turn that into a `std::unique_ptr<>`.

File: rtgui/adjuster.cc [hunk 6]
```diff
@@ -129,47 +128,41 @@ Adjuster::Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep
...
 }
 
 void Adjuster::addAutoButton (Glib::ustring tooltip)
```
Comment (Floessie) [inline rtgui/adjuster.cc]: `const&`.

File: rtgui/adjuster.cc [hunk 7]
```diff
@@ -317,32 +300,35 @@ void Adjuster::resetPressed (GdkEventButton* event)
...
 
 void Adjuster::setLimits (double vmin, double vmax, double vstep, double vdefault)
 {
+    sliderChange.block(true);
+    spinChange.block(true);
 
-    sliderChange.block (true);
-    spinChange.block (true);
+    double pow10 = vstep;
+    for (digits = 0; fabs(pow10 - floor(pow10)) > 0.000000000001; digits++, pow10 *= 10.0);
```
Comment (Floessie) [inline rtgui/adjuster.cc]: `std::fabs()` and friends? We include `<cmath>`, so everything should be prefixed with `std::`.

File: rtgui/adjuster.cc [hunk 8]
```diff
@@ -576,50 +555,38 @@ void Adjuster::showEditedCB ()
...
-        adjusterListener->adjusterChanged (this, spin->get_value ());
+        adjusterListener->adjusterChanged(this, spin->get_value());
     }
 
     eventPending = false;
 }
 
-void Adjuster::trimValue (double &val)
+void Adjuster::trimValue (double &val) const
 {
```
Comment (Floessie) [inline rtgui/adjuster.cc]: A lot of blank lines.

File: rtgui/adjuster.cc [hunk 9]
```diff
@@ -576,50 +555,38 @@ void Adjuster::showEditedCB ()
...
 
-    val = rtengine::LIM(val, static_cast<float>(vMin), static_cast<float>(vMax));
+    val = rtengine::LIM<float>(val, vMin, vMax);
 
 }
 
-
-inline double Adjuster::getSliderValue()
+inline double Adjuster::getSliderValue() const
```
Comment (Floessie) [inline rtgui/adjuster.cc]: Does `inline` have an effect here?

File: rtgui/adjuster.h [hunk 10]
```diff
@@ -75,28 +73,25 @@ class Adjuster : public Gtk::Grid
...
-    double getSliderValue();
+    double getSliderValue() const;
     void setSliderValue(double val);
 
 public:
 
     int delay;
 
-    Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep, double vdefault, Gtk::Image *imgIcon1 = nullptr, Gtk::Image *imgIcon2 = nullptr, double2double_fun slider2value = nullptr, double2double_fun value2slider = nullptr);
+    Adjuster (Glib::ustring vlabel, double vmin, double vmax, double vstep, double vdefault, Gtk::Image *imgIcon1 = nullptr, Gtk::Image *imgIcon2 = nullptr, double2double_fun slider2value_ = nullptr, double2double_fun value2slider_ = nullptr);
```
Comment (Floessie) [inline rtgui/adjuster.h]: Hmm, Alberto used a `_` suffix for members lately, which would clash. I prefer an underscore *prefix*.

File: rtgui/adjuster.h [hunk 11]
```diff
@@ -75,28 +73,25 @@ class Adjuster : public Gtk::Grid
...
 
     // Add an "Automatic" checkbox next to the reset button.
     void addAutoButton(Glib::ustring tooltip = "");
-    // Remove the "Automatic" checkbox next to the reset button.
-    void delAutoButton();
     // Send back the value of og the Auto checkbox
     bool getAutoValue ()
     {
-        return automatic != nullptr ? automatic->get_active () : false;
+        return automatic ? automatic->get_active() : false;
```
Comment (Floessie) [inline rtgui/adjuster.h]: Oh well, let's move implementations to the implementation file.

**Review by Floessie — CHANGES_REQUESTED** [review]
Well done, Ingo. There are only some small enhancements left.

File: rtgui/adjuster.cc (hunk 9 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Removed.

File: rtgui/adjuster.cc (hunk 8 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Removed

File: rtgui/adjuster.cc (hunk 5 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: I Gtk::manage(d) it for consistency

File: rtgui/adjuster.cc (hunk 6 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Done

File: rtgui/adjuster.cc (hunk 2 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: & added

File: rtgui/adjuster.cc (hunk 7 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Done

File: rtgui/adjuster.cc (hunk 4 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Done

File: rtgui/adjuster.h (hunk 11 above)
Comment (heckflosse) [reply rtgui/adjuster.h]: Done

File: rtgui/adjuster.h (hunk 10 above)
Comment (heckflosse) [reply rtgui/adjuster.h]: Changed

File: rtgui/adjuster.cc (hunk 1 above)
Comment (heckflosse) [reply rtgui/adjuster.cc]: Done, please have a look at how I did it. Dont know if we have a rule how to...

**Comment by heckflosse** [comment]
@Hombre57 

Maybe interesting also for you:

The speedup to create the adjusters was achieved by moving the 
```
    setLimits(vmin, vmax, vstep, vdefault);
```
call upwards before the
```
attach_next_to(...)
```
stuff

**Comment by Hombre57** [comment]
@heckflosse May I ask you, how did you analyzed / though about this speedup ? 😀

**Comment by heckflosse** [comment]
@Hombre57 

I placed a BENCFUNMCIRO in the adjuster constructor. That clearly told me each of the the adjuster construction time went down by 0.2 ms which multilied by ~300 adjusters (METM batch edit) is 60 ms and multiplied by ~600 (SETM) is 120 ms

**Comment by Hombre57** [comment]
@heckflosse Thanks for this tip, I'll learn how to use it.

File: rtgui/adjuster.cc [hunk 12]
```diff
@@ -25,93 +25,106 @@
...
+Adjuster::Adjuster (
+    Glib::ustring vlabel,
+    double vmin,
+    double vmax,
+    double vstep,
+    double vdefault,
+    Gtk::Image *imgIcon1,
+    Gtk::Image *imgIcon2,
+    double2double_fun slider2value,
+    double2double_fun value2slider)
```
Comment (Floessie) [inline rtgui/adjuster.cc:48]: You asked me about this. I usually break it this way, but yours is fine, too:

### C'tor
```c++
Class::Class(
    const Type& _param1,
    const Type& _param2,
    Enum _param3
) :
    param1(_param1),
    param2(_param2),
    param3(_param3)
{
}
```

### Function / method
```c++
ReturnType getReturnType(
    const Type& _param1,
    const Type& _param2,
    Enum _param3
)
{
    return {};
}
```

File: rtgui/adjuster.cc [hunk 13]
```diff
@@ -629,69 +602,125 @@ inline double Adjuster::getSliderValue()
...
     
-    sliderChange.block (false);
-    spinChange.block (false);
+    sliderChange.block(false);
+    spinChange.block(false);
+}
+
+bool Adjuster::getAutoValue() const
+{
+    return automatic ? automatic->get_active() : false;
```
Comment (Floessie) [inline rtgui/adjuster.cc:672]: Could also be `return automatic && automatic->get_active();`.

**Review by Floessie — APPROVED** [review]

**Pull request merged**